### PR TITLE
feat(scheduler-targets): SqsSendMessage Target

### LINKED ISSUE #27458: (aws-scheduler-targets-alpha): Add SqsSendMessage Target
### Describe the feature

Work to support L2 constructs for AWS Scheduler is in progress (https://github.com/aws/aws-cdk/issues/23394). See the approved [RFC](https://github.com/aws/aws-cdk-rfcs/pull/486/files). RFC planned to add 12 templates targets, but only Lambda Invoke is currently implemented (https://github.com/aws/aws-cdk/pull/26575). 

This issue tracks implementation of  `SqsSendMessage` target to send a message to an Amazon SQS Queue

### Use Case

Customers would like to use templated target `SqsSendMessage` to be able to send a message to an Amazon SQS Queue on schedule. L2 target construct should grant required permissions to the `AWS Scheduler` to send a message to an Amazon SQS Queue.

### Proposed Solution

The proposed solution needs to be adopted to the recent examples of `LambdaInvoke` (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-scheduler-targets-alpha/lib/lambda-invoke.ts). 

Solution should also include unit and integration tests. 

Class `SqsSendMessage` should: 
1. Grant Scheduler Execution Role permissions to send a message to an Amazon SQS Queue via `addTargetActionToRole`
2. Override `bindBaseTargetConfig` to return `sqsParameters`  as part of `ScheduleTargetConfig`.

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.99.1

### Environment details (OS name and version, etc.)

MacOS

### PR DESCRIPTION
**Pull request by go-to-k**
This PR adds `SqsSendMessage` Target for EventBridge Scheduler.

Closes #27458.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  constructor(
+    private readonly queue: sqs.IQueue,
+    private readonly props: SqsSendMessageProps,
+  ) {
+    super(props, queue.queueArn);
+
+    if (props.messageGroupId !== undefined) {
+      if (props.messageGroupId.length < 1 || props.messageGroupId.length > 128 ) {
+        throw new Error(`messageGroupId length must be between 1 and 128, got ${props.messageGroupId.length}`);
+      }
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:33]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-sqsparameters.html#cfn-scheduler-schedule-sqsparameters-messagegroupid

**Comment by go-to-k** [comment]
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-sqsparameters.html#cfn-scheduler-schedule-sqsparameters-messagegroupid

> Contains the message group ID to use when the target is a FIFO queue.

> If you specify an Amazon SQS FIFO queue as a target, the queue must have content-based deduplication enabled.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-target.html#cfn-scheduler-schedule-target-sqsparameters

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      if (!queue.fifo) {
+        throw new Error('target must be a FIFO queue if messageGroupId is specified');
+      }
+      if (!(queue.node.defaultChild as sqs.CfnQueue).contentBasedDeduplication) {
+        throw new Error('contentBasedDeduplication must be true if the target is a FIFO queue');
+      }
+    } else if (queue.fifo) {
+      throw new Error('messageGroupId must be specified if the target is a FIFO queue');
+    }
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:42]: > Contains the message group ID to use when the target is a FIFO queue.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-target.html#cfn-scheduler-schedule-target-sqsparameters

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    if (props.messageGroupId !== undefined) {
+      if (props.messageGroupId.length < 1 || props.messageGroupId.length > 128 ) {
+        throw new Error(`messageGroupId length must be between 1 and 128, got ${props.messageGroupId.length}`);
+      }
+      if (!queue.fifo) {
+        throw new Error('target must be a FIFO queue if messageGroupId is specified');
+      }
+      if (!(queue.node.defaultChild as sqs.CfnQueue).contentBasedDeduplication) {
+        throw new Error('contentBasedDeduplication must be true if the target is a FIFO queue');
+      }
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:39]: > If you specify an Amazon SQS FIFO queue as a target, the queue must have content-based deduplication enabled.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-target.html#cfn-scheduler-schedule-target-sqsparameters

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  ) {
+    super(props, queue.queueArn);
+
+    if (props.messageGroupId !== undefined) {
+      if (props.messageGroupId.length < 1 || props.messageGroupId.length > 128 ) {
+        throw new Error(`messageGroupId length must be between 1 and 128, got ${props.messageGroupId.length}`);
+      }
+      if (!queue.fifo) {
+        throw new Error('target must be a FIFO queue if messageGroupId is specified');
+      }
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:36]: CFn occurs a following error when the queue is not FIFO queue if messageGroupId is specified. So added this validation.

```
ROLLBACK_COMPLETE: Resource handler returned message: "Invalid request provided: Parameters MessageGroupId not valid for the target.
```

**Comment by go-to-k** [comment]
@kaizencc 

For some reason this PR doesn't seem to have a `pr/needs-community-review` label. How can I label it?

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (!sameEnvDimension(this.queue.env.account, schedule.env.account)) {
+      throw new Error(`Cannot assign queue in account ${this.queue.env.account} to the schedule ${Names.nodeUniqueId(schedule.node)} in account ${schedule.env.region}. Both the schedule and the queue must be in the same account.`);
+    }
+
+    if (this.props.role && !sameEnvDimension(this.props.role.env.account, this.queue.env.account)) {
+      throw new Error(`Cannot grant permission to execution role in account ${this.props.role.env.account} to invoke target ${Names.nodeUniqueId(this.queue.node)} in account ${this.queue.env.account}. Both the target and the execution role must be in the same account.`);
+    }
+
+    this.queue.grant(role, 'sqs:SendMessage');
```
Comment (kylelaker) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:65]:
```suggestion
    this.queue.grantSendMessages(role);
```

Would this be better? It handles the case where the queue uses KMS encryption and grants a few other permissions. But I'm not sure that's necessary for this case?

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  constructor(
+    private readonly queue: sqs.IQueue,
+    private readonly props: SqsSendMessageProps,
+  ) {
+    super(props, queue.queueArn);
+
+    if (props.messageGroupId !== undefined) {
+      if (props.messageGroupId.length < 1 || props.messageGroupId.length > 128 ) {
+        throw new Error(`messageGroupId length must be between 1 and 128, got ${props.messageGroupId.length}`);
+      }
```
Comment (kylelaker) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:40]: I don't think tokens ever get this long but maybe:

```suggestion
      if (!Token.isUnresolved(props.messageGroupId) && (props.messageGroupId.length < 1 || props.messageGroupId.length > 128)) {
        throw new Error(`messageGroupId length must be between 1 and 128, got ${props.messageGroupId.length}`);
      }
```

**Review by kylelaker — COMMENTED** [review]
The links to documentation are super helpful! Thanks! The tests also seem to cover the cases well.

I do have two small thoughts I wanted to ask about though.

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts (hunk 5 above)
Comment (go-to-k) [reply packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:65]: Certainly, I'll leave it at that.

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts (hunk 6 above)
Comment (go-to-k) [reply packages/@aws-cdk/aws-scheduler-targets-alpha/lib/sqs-send-message.ts:40]: Yes, I'll do that just in case.

**Comment by go-to-k** [comment]
@kylelaker 

Thanks for your review!

I changed, so please review again.

**Review by kylelaker — APPROVED** [review]
Thank you!

**Review by sumupitchayan — APPROVED** [review]
Thanks for contributing!

**Review by sumupitchayan — DISMISSED** [review]

**Review by sumupitchayan — APPROVED** [review]

**Pull request merged**